ast/index_test.go

```diff
@@ -912,3 +912,369 @@ func TestSkipIndexing(t *testing.T) {
 		t.Fatalf("Expected else to be %v but got: %v", expectedElse[r0], result.Else[r0])
 	}
 }
+
+func TestBaseDocIndexResultEarlyExit(t *testing.T) {
+
+	tests := []struct {
+		note            string
+		module          *Module
+		input           string
+		disableIndexing bool
+		expectedRS      interface{}
+		expectedDR      *Rule
+		expectedEE      bool
+	}{
+		{
+			note:       "single rule",
+			expectedEE: true,
+			module: MustParseModule(`package test
+r {
+	input.x = 1
+}
+r = 3 {
+	input.y = 2
+}`),
+			input: `{"x": 1}`,
+			expectedRS: []string{
+				`r { input.x = 1 }`,
+			},
+		},
+		{
+			note:            "no early exit: two rules, indexing disabled",
+			disableIndexing: true,
+			expectedEE:      false,
+			module: MustParseModule(`package test
+r {
+	input.x = 1
+}
+r = 3 {
+	input.y = 2
+}`),
+			input: `{"x": 1}`,
+			expectedRS: []string{
+				`r = 3 { input.y = 2 }`,
+				`r { input.x = 1 }`,
+			},
+		},
+		{
+			note:            "two rules, indexing disabled",
+			disableIndexing: true,
+			expectedEE:      true,
+			module: MustParseModule(`package test
+r {
+	input.x = 1
+}
+r {
+	input.y = 2
+}`),
+			input: `{"x": 1}`,
+			expectedRS: []string{
+				`r { input.y = 2 }`,
+				`r { input.x = 1 }`,
+			},
+		},
+		{
+			note:       "no early exit: different constant value",
+			expectedEE: false,
+			module: MustParseModule(`package test
+r {
+	input.x = 1
+}
+r = 2 {
+	input.x = 1
+	input.y = 2
+}`),
+			input: `{"x": 1, "y": 2}`,
+			expectedRS: []string{
+				`r { input.x = 1 }`,
+				`r = 2 { input.x = 1; input.y = 2 }`,
+			},
+		},
+		{
+			note:       "same constant value",
+			expectedEE: true,
+			module: MustParseModule(`package test
+r {
+	input.x = 1
+}
+r {
+	input.y = 1
+}`),
+			input: `{"x": 1, "y": 1}`,
+		},
+		{
+			note:       "no early exit: one rule with with non-constant value",
+			expectedEE: false,
+			module: MustParseModule(`package test
+r {
+	input.x = 1
+}
+r = x {
+	input.y = 1
+	x = "foo"
+}`),
+			input: `{"x": 1, "y": 1}`,
+			expectedRS: []string{
+				`r { input.x = 1 }`,
+				`r = x { input.y = 1; x = "foo" }`,
+			},
+		},
+		{
+			note:       "same ref value (input)",
+			expectedEE: true,
+			module: MustParseModule(`package test
+r = input.a {
+	input.x = 1
+}
+r = input.a {
+	input.y = 1
+}`),
+			input: `{"x": 1, "y": 1}`,
+		},
```

Comment:
@tsandall Your experiment's branch used `ast.IsConstant` and this excluded this case and the next -- Was that on purpose? I might be missing something 🤔

Reply:
@srenatus sorry, no, that wasn't intentional--I just made the indexer changes to test out the alternative EE implementation in topdown. The indexer changes weren't intended be shipped.

re: IsGround vs. IsConstant and ground refs in the head:

The compiler will actually rewrite rules that contain refs in the head so that the head only contains vars and those vars are bound to refs in the body--so I don't think it will matter.

Reply:
Ah alright. Fooled myself with the test cases once again, I suppose. 😅

Reply:
Added a note to the test cases. Removing them would have gotten me wondering for a bit next year or so, I know me.